dessia_common/workflow/core.py

```diff
@@ -508,54 +504,51 @@ def __deepcopy__(self, memo=None):
         return copied_workflow
 
     def copy_pipe(self, copied_workflow: 'Workflow', pipe: Pipe) -> Pipe:
+        """ Copy a single regular pipe. """
         upstream_index = self.variable_indices(pipe.input_variable)
-        if isinstance(upstream_index, int):
+        if self.is_variable_nbv(pipe.input_variable):
             raise dessia_common.errors.CopyError("copy_pipe method cannot handle nonblock-variables. "
-                                                 "Please consider using copy_pipes")
+                                                 "Please consider using copy_nbv_pipes")
         pipe_upstream = copied_workflow.variable_from_index(upstream_index)
 
         downstream_index = self.variable_indices(pipe.output_variable)
         pipe_downstream = copied_workflow.variable_from_index(downstream_index)
         return Pipe(pipe_upstream, pipe_downstream)
 
-    def copy_pipes(self, copied_workflow: 'Workflow') -> List[Pipe]:
-        copied_pipes = []
-
-        nbv_pipes = []
-        standard_pipes = []
-        for pipe in self.pipes:
-            upstream_index = self.variable_indices(pipe.input_variable)
-            if isinstance(upstream_index, int):
-                nbv_pipes.append(pipe)
-            else:
-                standard_pipes.append(pipe)
-
-        for pipe in standard_pipes:
-            copied_pipes.append(self.copy_pipe(copied_workflow, pipe))
+    def copy_nbv_pipe(self, copied_workflow: 'Workflow', pipe: Pipe, copy_memo: Dict[int, Variable]) -> Pipe:
+        """
+        Copy a pipe where its upstream variable is a NBV.
 
-        for nbv in self.nonblock_variables:
-            related_pipes = [pipe for pipe in nbv_pipes if pipe.input_variable == nbv]
-            if related_pipes:
-                copied_variable = nbv.copy()
-                for pipe in related_pipes:
-                    downstream_index = self.variable_indices(pipe.output_variable)
-                    pipe_downstream = copied_workflow.variable_from_index(downstream_index)
-                    copied_pipes.append(Pipe(copied_variable, pipe_downstream))
+        This needs special care because if it is not handled properly, NBVs can duplicate,
+        or copied pipes might be unordered.
+        """
+        nbv = pipe.input_variable
+        upstream_index = self.variable_index(nbv)
+        if upstream_index in copy_memo:
+            copied_variable = copy_memo[upstream_index]
+        else:
+            copied_variable = nbv.copy()
+            copy_memo[upstream_index] = copied_variable
```

Comment:
can upstream_index be None ?
If yes, what if no value in upstream_index ?

Reply:
I'll have to give it a second thought. I see no way it could be None, but now that you mention it, I have to ensure that it cannot